New articles for /payments section (#3472)
GoogleChrome/web.dev: src/site/content/en/payments/registering-a-web-based-payment-app/index.md

---
layout: post
title: Registering a web-based payment app
subhead: |

>   Learn how to configure a web-based payment app on registration.

Reviewer: ```suggestion
  Learn how to configure a web-based payment app during service worker registration.
```

Author: This "registration" is for a web-based payment app. Not necessarily about a service worker.

workers](https://developer.mozilla.org/docs/Web/API/Service_Worker_API). The
service worker in a payment app plays an important role as it captures payment
requests from a merchant, launches the payment app, and mediates the
communication with the merchant.

To configure a web-based payment app, you need to register available payment
methods, delegation options, and a service worker. You can configure your
web-based payment app declaratively with a web app manifest or imperatively with
JavaScript.

## Configuring a payment app with a web app manifest

To configure your web-based payment app declaratively, [serve a web app
manifest](/setting-up-a-payment-method/#step-3:-serve-a-web-app-manifest).

The following properties in the web app manifest are relevant for web-based payment apps:
* `name`
* `icons`
* `serviceworker`
    * `src`
    * `scope`
    * `use_cache`
* `payment`
    * `supported_delegations`

Check out [Setting up a payment
method](/setting-up-a-payment-method/#step-3:-serve-a-web-app-manifest)
to make sure your payment method manifest points to your web app manifest
properly.

### Registering a service worker just-in-time (JIT)

> Service workers are usually registered using JavaScript, but they can also be
> automatically registered by the browser when the user chooses to pay with a
> web-based payment app on a merchant website. This is called "just-in-time (JIT)
> registration".

Reviewer: ```suggestion
{% Aside 'key-term' %}
Service workers are usually registered using JavaScript, but they can also be
automatically registered by the browser when the user chooses to pay with a
web-based payment app on a merchant website. This is called *just-in-time (JIT)
registration*.
{% endAside %}
```
https://web.dev/handbook/web-dev-components/#key-term-asides

The JIT registration requires only serving [the web app
manifest](/setting-up-a-payment-method/#step-3:-serve-a-web-app-manifest)
and no additional coding. If you've already configured your web app manifest and
are serving it properly, you should be all set. The browser will handle the

> rest.

Reviewer: All browsers support JIT registration?

Author: It's not defined as the standard behavior, but the Payment Handler API is only available in Chrome currently anyway.

{% Aside 'caution' %}
While a single payment method identifier can support multiple payment apps,

> JIT-registration happens only when the payment method manifest points to a

Reviewer: ```suggestion
JIT registration happens only when the payment method manifest points to a
```

single payment app.
{% endAside %}

## Configure a payment app with JavaScript

Configuring a web-based payment app using JavaScript allows you to register
multiple payment methods and manually unregister service workers.

> ### Manually register a service worker

Reviewer: Let's recommend using Workbox somewhere in this section?

Maybe we can use this include: https://github.com/GoogleChrome/web.dev/blob/master/src/site/_includes/content/reliable/workbox.njk

You would use this code to add the include: `{% include 'content/reliable/workbox.njk' %}`

Author: Why? Workbox doesn't have anything that helps payment handler?

Reviewer: If someone only uses a service worker to handle Payment Handler stuff, then I agree, Workbox is unnecessary.

However, I'm worried that once we make our audience aware of service workers, they will realize that they can use it for other tasks, such as caching. In that case, we absolutely want them to consider Workbox first.

So, how about we rephrase the section along these general lines:

* If you're **only** using a service worker for basic Payment Handler logic, then you can manually implement one yourself.
* If you start using your service worker for other use cases, such as caching, we strongly recommend using Workbox as the foundation of your service worker.

cc @jeffposnick

{% Aside %}
Learn more about service worker registration in [Service Workers: an
Introduction](https://developers.google.com/web/fundamentals/primers/service-workers).
{% endAside %}

### Set a payment instrument

{% Aside 'key-term' %}
Payment apps can support multiple ways for making a payment. For example, a
customer can add multiple credit cards or e-wallets to a payment app. Each of
them is a payment instrument.
{% endAside %}

Once the service worker is registered, [a service worker registration
object](https://developer.mozilla.org/docs/Web/API/ServiceWorkerRegistration) is
returned. Call
[`registration.paymentManager.instrument.set()`](https://w3c.github.io/payment-handler/#paymentinstruments-interface)
to set a payment instrument. It accepts two arguments, a string that represents
the instrument key and a [`PaymentInstrument`
object](https://w3c.github.io/payment-handler/#dom-paymentinstrument) that
contains details about the instrument.

{% Label %}[payment handler] payment app's landing pages{% endLabel %}

```js
...
  // Feature detection
  if (!registration.paymentManager) return;

  await registration.paymentManager.instruments.set(
    // Payment instrument key

>     "bobpay_card_1",

Reviewer: ```suggestion
    'bobpay_card_1',
```
Please check all code samples to make sure they use single quotes

    // Payment instrument details
    {
      // This parameter will be ignored in Chrome
      name: 'Payment Handler Example',
      // This parameter will be used to match against
      // the PaymentRequest.
      method: 'https://bobpay.xyz/pay'
    }
  );
```

> <div class="w-table-wrapper">

Reviewer: @robdodson this nested table looks a bit unorganized to me, can you think of any other way to represent it or make it look a bit more organized?

Reviewer: @agektmr I will push up a commit proposing a way to make this table less complicated. I'm worried that it won't display correctly on mobile.

        </td>
      </tr>
      <tr>
        <td>
          <a href="https://developer.mozilla.org/docs/Web/Manifest/icons"><code>icons</code></a>
        </td>
        <td>
          Sets an array of <a href="https://w3c.github.io/payment-handler/#imageobject-dictionary"><code>ImageObject</code></a>s that the Payment Request sheet will display. Ignored in Chrome.
        </td>
      </tr>
      <tr>
        <td>
          <code>method</code>
        </td>
        <td>
          A supported payment method identifier.
        </td>
      </tr>
      <tr>
        <td>
          <code>capabilities</code>
        </td>
        <td>
          Sets the <a href="https://w3c.github.io/payment-handler/#capabilities-example">payment method specific parameters as an object</a>. As of July 2020, basic-card is the only payment method that accepts capabilities.
        </td>
      </tr>
    </tbody>
  </table>
</div>

{% Aside %}
Chrome ignores the `name` and `icons` properties. It respects the [web app
manifest](/setting-up-a-payment-method/#step-3:-serve-a-web-app-manifest)'s
respective properties instead, but other browsers may behave differently.

> {% endAside %}

Reviewer: We don't know how other browsers behave? As a web developer considering Web Payments, I think the story will be much more compelling if this article at least tells me the current state of things in other browsers

Author: Again, this API is not implemented in any other browsers yet. Though Firefox just started doing so.

### Enable delegations

Entering shipping address and contact information through a web form can be a
cumbersome experience for customers. It can cause errors and lower conversion
rate.

That's why the Payment Request API supports a feature to request shipping
address and contact information. This provides multiple benefits:

* Users can pick the right address with just a few taps.
* The address is always returned in [the standardized
  format](https://w3c.github.io/payment-request/#paymentaddress-interface).
* Submitting an incorrect address is less likely.

> Payment apps can integrate this functionality to offer a more unified payment
> experience and this is called "delegation".

Reviewer: ```suggestion
Payment apps can integrate this functionality to offer a more unified payment
experience. This is called *delegation*.
```

<figure class="w-figure" style="width:300px; margin:auto;">
  <video controls autoplay loop muted class="w-screenshot">
    <source src="https://storage.googleapis.com/web-dev-assets/payments/skip-the-sheet.webm" type="video/webm">
    <source src="https://storage.googleapis.com/web-dev-assets/payments/skip-the-sheet.mp4" type="video/mp4">
  </video>
  <figcaption class="w-figcaption">
    Checkout flow with a web-based payment app.
  </figcaption>
</figure>

Web-based payment apps can declare their support for delegation using the web
app manifest's `payment.supported_delegations` field or through JavaScript.

To let the browser know that the payment app accepts a delegation, use

> [`.enableDelegations()`](https://w3c.github.io/payment-handler/#enabledelegations-method).

Reviewer: ```suggestion
[`PaymentManager.enableDelegations()`](https://w3c.github.io/payment-handler/#enabledelegations-method).
```
Let's follow MDN's convention. They include the name of the class/object before the method name. Example: https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API#Registration

Author: Good suggestion. I'll do so moving forward.

{% Label %}[payment handler] payment app's landing page{% endLabel %}

```js

> ...
>  await registration.paymentManager.enableDelegations([
>   'shippingAddress', 'payerName'
> ]);
> ...

Reviewer: ```suggestion
 await registration.paymentManager.enableDelegations([
  'shippingAddress', 'payerName'
]);
```
Let's try getting rid of these ellipses because it seems to be messing up the syntax highlighting

Author: Hmm, I'm trying to express that there's hidden code before and after this snippet. Any alternative ideas?

```

You can declare supported delegation options as an array of strings:

* `shippingAddress`: The payment handler can provide a shipping address.
* `payerName`: The payment handler can provide the payer's name.
* `payerPhone`: The payment handler can provide the payer's phone number.
* `payerEmail`: The payment handler can provide the payer's address.

## Debugging a web-based payment app

When developing a web-based payment app frontend, you'll probably jump between

> merchant perspective and payment app perspective. The following debugging tips

Reviewer: Perspective?

Author: The merchant and the payment app are two separate web contexts the developer has to deal with. If there' any better wording on this, please let me know.

Reviewer: `context`?

will help your developing experience on Chrome.

### Developing on a local server

Which server do you use for development? Many developers tend to use localhost
or a company-internal server environment which can be challenging because
powerful features in the browser tend to require a secure environment (HTTPS)
and a valid certificate. The Payment Request API and the Payment Handler API are
no exception and localhosts or company-internal servers usually don't come with
a valid certificate.

The good news is some browsers, including Chrome, exempt certificates for
`http://localhost` by default. Also in Chrome, you can exempt the certificate
requirement by launching a Chrome instance. For example, to exempt the
requirement from `http://*.corp.company.com`, use the following flags:

* [`--ignore-certificate-errors`](https://chromiumdash.appspot.com/commit/988b56b519836f3d3d94f145ba3554a0c0a7d0a8)
* [`--unsafely-treat-insecure-origin-as-secure=http://*.corp.company.com`](https://chromiumdash.appspot.com/commit/77a7e1a65d14072149ec4420a0ab523586011d8a)

> 

Reviewer: Please provide a full example of launching Chrome from the command line. Ideally on Mac and Windows.

### Port forwarding a local server

You can port forward the local web server to an Android device using Chrome's
DevTools and test how it works from a mobile browser. To learn how to do it,
check out [Access Local

> Servers](https://developers.google.com/web/tools/chrome-devtools/remote-debugging/local-server).

Reviewer: FYI I haven't looked at that guide in a long time and I'm pretty sure some developers complained that the workflow didn't work. Have you tried it yourself?

Author: No, I didn't :/

Reviewer: I know this is extra work but can you try it out? I would do it myself but you're way more familiar with these APIs and should be able to do it faster. It's really disrespectful of our users if we recommend debugging workflows without knowing whether they actually work.

Author: https://github.com/google/WebFundamentals/pull/8876

### Remote debugging a website on Android Chrome from desktop DevTools

You can also debug Android Chrome on desktop DevTools. To learn how to do it,
check out [Get Started with Remote Debugging Android
Devices](https://developers.google.com/web/tools/chrome-devtools/remote-debugging).

### Payment Handler event logging

DevTools can display Payment Handler API events for easier local development.

> Open DevTools on the merchant context and go to the "Payment Handler" section

Reviewer: If I recall correctly, I think you need to first click that **Record** button?

We should link to the announcement: https://developers.google.com/web/updates/2019/09/devtools#payments

> Open DevTools on the merchant context and go to the "Payment Handler" section
> under the "Application" pane.

Reviewer: ```suggestion
Open DevTools on the merchant context and go to the **Payment Handler** section
under the **Application** pane.
```

https://developers.google.com/style/ui-elements#formatting

> ![Payment Handler event logging](debug-payment-handler.png)

Reviewer: Please format this as a full screenshot: https://web.dev/handbook/markup-media/#entire-screenshots

You can capture events sent to the service worker that handles payments.

> {% Aside 'caution' %}
> Make sure that "Show events from other domains" is enabled.
> {% endAside %}

Reviewer: ```suggestion
```
This is important information for the debugging workflow and therefore shouldn't be in an aside. Please add it into the paragraph above where you provide the debugging workflow instructions

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```

You learned how to register a service worker, set payment instruments, and
define delegation availability for a web-based payment app. The next step is to
learn how the service worker can orchestrate a payment transaction at runtime.

> * Orchestrating the payment transaction on the service worker

Reviewer: Are you still working on this guide? Will it be added to this PR? Or another PR?

Author: I have two more articles to be ready soon, but will create a separate pull request.
